pkg/mic/mic.go

```diff
@@ -1317,3 +1327,117 @@ func (c *Client) checkIfIdentityImmutable(id string) bool {
 	}
 	return false
 }
+
+// generateIdentityAssignmentState generates the current and desired state of each node's identity
+// assignments based on an existing list of AzureAssignedIdentity as the source of truth.
+func (c *Client) generateIdentityAssignmentState() (currentState map[string]map[string]bool, desiredState map[string][]string, isVMSSMap map[string]bool, err error) {
+	type nodeMetadata struct {
+		nodeNameOnAzure string
+		isVMSS          bool
+	}
+
+	assignedIDs, err := c.CRDClient.ListAssignedIDs()
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("failed to list AzureAssignedIdentities, error: %+v", err)
+	}
+
+	nodeMetadataCache := make(map[string]nodeMetadata)
+	isVMSSMap = make(map[string]bool)
+	currentState = make(map[string]map[string]bool)
+	desiredState = make(map[string][]string)
+	for _, assignedID := range *assignedIDs {
+		if _, ok := nodeMetadataCache[assignedID.Spec.NodeName]; !ok {
+			node, err := c.NodeClient.Get(assignedID.Spec.NodeName)
+			if err != nil {
+				return nil, nil, nil, fmt.Errorf("failed to get node %s, error %+v", assignedID.Spec.NodeName, err)
+			}
+
+			nodeNameOnAzure, isVMSS, err := isVMSS(node)
+			if err != nil {
+				return nil, nil, nil, fmt.Errorf("failed to check if node %s is VMSS, error: %+v", assignedID.Spec.NodeName, err)
+			} else if isVMSS {
+				nodeNameOnAzure = getVMSSName(nodeNameOnAzure)
+			} else {
+				// VM node name does not require conversion
+				nodeNameOnAzure = assignedID.Spec.NodeName
+			}
+
+			// cache node metadata to avoid excessive GET calls
+			nodeMetadataCache[assignedID.Spec.NodeName] = nodeMetadata{
+				nodeNameOnAzure: nodeNameOnAzure,
+				isVMSS:          isVMSS,
+			}
+		}
+
+		nodeNameOnAzure := nodeMetadataCache[assignedID.Spec.NodeName].nodeNameOnAzure
+		isVMSS := nodeMetadataCache[assignedID.Spec.NodeName].isVMSS
+		isVMSSMap[nodeNameOnAzure] = isVMSS
+
+		// only consider AzureAssignedIdentities in ASSIGNED state
+		// do not consider AzureAssignedIdentities in CREATED state because they are either:
+		// 1. in the process of assigning the identities on Azure or
+		// 2. encountering errors when assigning identities on Azure
+		if assignedID.Status.Status == aadpodid.AssignedIDAssigned {
```

Comment:
We also need to check if this `AzureIdentity` is of type User MSI. If not, then we should skip as it doesn't contain a `ResourceID`.